Source/WebCore/page/Navigation.cpp

```diff
@@ -717,48 +763,73 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
 
     if (endResultIsSameDocument) {
         Vector<RefPtr<DOMPromise>> promiseList;
-        bool failure = false;
 
         for (auto& handler : event->handlers()) {
             auto callbackResult = handler->handleEvent();
             if (callbackResult.type() == CallbackResultType::Success)
                 promiseList.append(callbackResult.releaseReturnValue());
-            else
-                failure = true;
-            // FIXME: We need to keep around the failure reason but the generated handleEvent() catches and consumes it.
+            else if (callbackResult.type() == CallbackResultType::ExceptionThrown) {
+                // FIXME: We need to keep around the failure reason but the generated handleEvent() catches and consumes it.
+                auto promiseAndWrapper = createPromiseAndWrapper(*document);
+                promiseAndWrapper.second->reject(ExceptionCode::TypeError);
+                promiseList.append(WTFMove(promiseAndWrapper.first));
+            }
         }
 
-        // FIXME: Step 33.4: We need to wait for all promises.
+        if (promiseList.isEmpty()) {
+            auto promiseAndWrapper = createPromiseAndWrapper(*document);
+            promiseAndWrapper.second->resolveWithCallback([](JSDOMGlobalObject&) {
+                return JSC::jsUndefined();
+            });
+            promiseList.append(WTFMove(promiseAndWrapper.first));
+        }
 
-        if (document->isFullyActive() && !abortController->signal().aborted()) {
-            // If a new event has been dispatched in our event handler then we were aborted above.
-            if (m_ongoingNavigateEvent != event.ptr())
-                return false;
+        waitForAllPromises(promiseList, [=, this, weakThis = WeakPtr { *this }]() mutable {
+            if (!weakThis || abortController->signal().aborted() || !document->isFullyActive() || !m_ongoingNavigateEvent)
```

Comment:
Can we make a RefPtr strongThis = weakThis.get()?  That'll prevent a callback destroying the Navigation object somehow from hitting a UAF with the raw this pointer.  Same below.